src/katgpucbf/fgpu/kernels/ddc.mako

```diff
@@ -14,307 +14,234 @@
-<%namespace name="wg_reduce" file="/wg_reduce.mako"/>
-
-/* TODO:
- * - test reading via shared memory with memcpy_async
- * - test writing via shared memory
- * - test putting constants into shared memory
- * - look into __constant memory type
- * - Try packing multiple logical workgroups into one physical one e.g. to
- *   enable warp-synchronous operation without harming occupancy
- * - Check if it could be more efficient to load SG_SIZE strided values per
- *   tile, instead of contiguous
- */
-
 /* Alignment requirements:
- * - in_offset must be a multiple of SEGMENT_SAMPLES
- * - GROUP_IN_SIZE must be a multiple of SEGMENT_SAMPLES
- * - SEGMENT_SAMPLES * SAMPLE_BITS must be a multiple of 32
- * - WGS must be a multiple of SG_SIZE
- * - TAPS must be a multiple of SUBSAMPLING
- * - SEGMENT_SAMPLES must be a multiple of SG_SIZE
- * - SUBSAMPLING must be a multiple of SG_SIZE
- * - COARSEN should be odd for best performance
- * - SG_SIZE should be a power of two for best performance
+ * - C * SUBSAMPLING * INPUT_SAMPLE_BITS must be a multiple of SAMPLE_WORD_BITS
  */
 
+<%include file="/port.mako"/>
+
 #define WGS ${wgs}
 #define TAPS ${taps}
 #define SUBSAMPLING ${subsampling}
-#define COARSEN ${coarsen}
-#define SG_SIZE ${sg_size}
-#define SAMPLE_BITS ${sample_bits}
-#define SEGMENT_SAMPLES ${segment_samples}
+#define INPUT_SAMPLE_BITS ${input_sample_bits}
+#define C ${unroll}
+#define W ${(taps + subsampling - 1) // subsampling}
 
-// Number of contiguous 32-bit words to store in each segment
-#define SEGMENT_WORDS (SEGMENT_SAMPLES * SAMPLE_BITS / 32)
-// Number of output samples
-#define GROUP_OUT_SIZE (COARSEN * (WGS / SG_SIZE))
-// Stride of input samples between workgroups
-#define GROUP_IN_SIZE (GROUP_OUT_SIZE * SUBSAMPLING)
-// Number of input samples to load in this workgroup
-#define LOAD_SIZE (GROUP_IN_SIZE + TAPS - SUBSAMPLING)
-// Load-size expressed in 32-bit words (rounding up)
-#define LOAD_WORDS ((LOAD_SIZE - 1) * SAMPLE_BITS / 32 + 1)
-// Number of segments to load per work-item
-#define SEGMENTS ((LOAD_SIZE - 1) / (SEGMENT_SAMPLES * WGS) + 1)
-/* Number of contiguous samples that take turns occupying a tile
- * (must divide both SEGMENT_SAMPLES and SUBSAMPLING, and be a multiple
- * of SG_SIZE). This implementation is pessimistic when SUBSAMPLING is
- * neither a factor nor multiple of SEGMENT_SAMPLES, but that's not expected to
- * be a common case.
- */
-#if SUBSAMPLING % SEGMENT_SAMPLES == 0
-# define TILE_SAMPLES SEGMENT_SAMPLES
-#elif SEGMENT_SAMPLES % SUBSAMPLING == 0
-# define TILE_SAMPLES SUBSAMPLING
-#else
-# define TILE_SAMPLES SG_SIZE
-#endif
-// Number of tiles to store in local memory
-#define TILES (LOAD_SIZE / TILE_SAMPLES)
-#define TILES_PER_SEGMENT (SEGMENT_SAMPLES / TILE_SAMPLES)
-#define TILES_PER_SUBSAMPLING (SUBSAMPLING / TILE_SAMPLES)
+/// Raw storage type for sample data
+typedef unsigned int sample_word;
+/// Signed version of sample_word
+typedef int ssample_word;
+#define SAMPLE_WORD_BITS 32
```

Comment:
I'm rusty on my C++ / CUDA but is `int` guaranteed to be 32-bits?

Reply:
Not in C++ in general, but it always is in CUDA. I'd much prefer to use `int32_t` to be explicit, but CUDA doesn't provide it.